flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -525,63 +525,88 @@ class ExprLowering {
   }
 
   template <Fortran::common::TypeCategory TC, int KIND>
-  mlir::Value
-  genval(const Fortran::evaluate::Constant<Fortran::evaluate::Type<TC, KIND>>
-             &con) {
-    // TODO:
-    // - array constant not handled
-    // - derived type constant
+  mlir::Value genScalarLit(
+      const Fortran::evaluate::Scalar<Fortran::evaluate::Type<TC, KIND>>
+          &value) {
     if constexpr (TC == Fortran::lower::IntegerCat) {
-      auto opt = con.GetScalarValue();
-      if (opt.has_value())
-        return genIntegerConstant<KIND>(builder.getContext(), opt->ToInt64());
-      llvm_unreachable("integer constant has no value");
+      return genIntegerConstant<KIND>(builder.getContext(), value.ToInt64());
     } else if constexpr (TC == Fortran::lower::LogicalCat) {
-      auto opt = con.GetScalarValue();
-      if (opt.has_value())
-        return genLogicalConstantAsI1(builder.getContext(), opt->IsTrue());
-      llvm_unreachable("logical constant has no value");
+      return genLogicalConstantAsI1(builder.getContext(), value.IsTrue());
```

Comment:
Why is a logical constant an I1?  My understanding is that I1 is a boolean type. Logical is not boolean.

Reply:
This whole function is just refactored code, I don't know the answer to your question. Maybe @jeanPerier @schweitzpgi can help me out here.

Reply:
Coincidentally, I just renamed this in another PR.